### PR TITLE
Update Docker file with multistage build and streamlining

### PR DESCRIPTION
**Pull request by marcpomfret**
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

### CODE REVIEW

**Comment by marcpomfret** [comment]
This PR streamlines the Collections image by installing dependencies etc in a new intermediate builder stage, before copying back the build outputs to the output image.
Please see previous commit for more details: alphagov/publisher#1519

A few values have been removed such as PORT  - These are not needed as part of the Build stage.

Adding Puma, as we want to run Puma in production, we are instead configuring it via govuk_app_config

File: Dockerfile [hunk 1]
```diff
@@ -1,22 +1,34 @@
...
+RUN bundle config set deployment 'true' && \
+    bundle config set without 'development test' && \
+    bundle install --jobs 4 --retry=2
+COPY . /app
 
-ADD . $APP_HOME
+RUN GOVUK_APP_DOMAIN=www.gov.uk \
+    GOVUK_WEBSITE_ROOT=http://www.gov.uk \
+    bundle exec rails assets:precompile
+# QQ:  rake shared_mustache:compile is this app specific?
```
Comment (bilbof) [inline Dockerfile]: ~This line can be removed. AFAIK collections does not use the mustache templating library.~

Ah I see you removed this in a later commit.

**Review by bilbof — COMMENTED** [review]
Looks great, thank you for working on this! 🌟 

Nit: please squash your commits so that your later commits aren't "cleaning up" after the earlier commits. E.g. the commit "new line in puma.rb" should be squashed into the previous commit "update puma config". We have a high bar for commit messages - I'll try to find our styleguide and share it with the group.

**Review by bilbof — APPROVED** [review]

**Comment by chao-xian** [comment]
Here are the guidelines Bill mentioned:

* https://gds-way.cloudapps.digital/standards/source-code.html#commit-messages
* https://gds-way.cloudapps.digital/standards/pull-requests.html

Particularly good read: https://cbea.ms/git-commit/

**Pull request merged**